logstash-core/src/main/java/org/logstash/FieldReference.java

```diff
@@ -179,4 +172,79 @@ private static FieldReference parse(final CharSequence reference) {
             return new FieldReference(path.toArray(EMPTY_STRING_ARRAY), key, DATA_CHILD);
         }
     }
+
+    /**
+     * The {@link StrictTokenizer} parses field-references in a strict manner; when illegal syntax is encountered,
+     * the input is considered ambiguous and the reference is not expanded.
+     **/
+    private static class StrictTokenizer {
+
+        public List<String> tokenize(CharSequence reference) {
+            ArrayList<String> path = new ArrayList<>();
+            final int length = reference.length();
+
+            boolean potentiallyAmbiguousSyntaxDetected = false;
+            int splitPoint = 0;
+            int depth = 0;
+            scan: for (int i=0 ; i < length; i++) {
```

Comment:
I'm of the opinion that labels for control flow are undesirable in Java code. Maybe see what @andrewvc or @original-brownbear think about them?

Reply:
I would agree that this kind of code presents the strongest case for labels. Unless others have stronger opinions on it, I'm ok with it here.

Reply:
@danhermann @yaauie yea I'm not a big fan of labels in control flow either. I'd rather not introduce them now, but we can refactor later I guess if it's a big hassle to clean that up now.